rfc-176-replacing-static.md

```diff
@@ -0,0 +1,127 @@
+
+Asset pre-compilation happens when the container image is built (during the ‘deploy’ GitHub Actions workflow). This process generates the compiled assets and the Sprockets manifest. The manifest is the mechanism for Rails to map the original asset name (`crown.png`) to the fingerprinted asset name (`crown-b2ba232ade53223.png`), so the deployed app has a record of everything that was compiled as part of its assets:precompile step.
+
+Upon deployment into the Kubernetes cluster, a job runs which copies anything from `/app/public/assets/<name of app>` into an S3 bucket `s3://govuk-app-assets-<environment>/assets/<name of app>`. [Fastly maps] any call for (e.g.) `https://www.gov.uk/assets/` to that bucket, so a request for `https://www.gov.uk/assets/frontend/crown-b2ba232ade53223.png` will be served from that s3 bucket.
+
+
+![Current asset deploy pathway](rfc-176/current_govuk_asset_pathway.png)
+
+#### What this means for shared assets
+Because the upload-assets job in [govuk-helm-charts] includes the app name, there currently isn’t a mechanism for apps to upload assets in a shared context. If two apps both precompile a CSS file for the Table component, for instance, the fingerprint will be the same, but the path will be different, so a client that views two pages with Tables on (one in [frontend], one in [government-frontend]) will download the same file twice.
+
+However, because all of these apps upload to different folders in the same S3 bucket, it would be relatively easy to add a second step to the upload which took assets from (say) `/app/public/assets/shared` on the deployed images and upload that to a shared folder. Because the same asset compiled in two different apps ends up with the same name, the second app’s uploaded asset would overwrite the first one with the exact same name and data. As long as mechanisms were in place to prevent confusion from two apps uploading at the same time (this may already be present in S3), we could safely write to the same shared assets directory.
+
+#### How will we do this in practice?
+Although sprockets-rails wants to put things in one directory (the assets.prefix directory configured in application.rb), it is possible to split assets by source by modifying the Sprockets::Asset class to return a digest path with different prefixes based on the asset name. Since asset engine gems tend to be namespaced (govuk_publishing_components is, for instance), we can use the gem name to determine whether a thing should go into shared or not. If the asset starts with a gem name on a list we provide, it can be prefixed with /shared, and if not it can be prefixed with the application name as usual.
```

Comment:
Yeah, I think I was seduced by the fact that we currently have separate folders into thinking that that was a necessary solution. Now that you've mentioned it, I think you're right. It's not important to have segregation of shared and individual assets in the final buckets, because we shouldn't be doing anything with those buckets other than adding to them (and tidying them, but tidying isn't made much easier by segregating the folders).

The simplest thing would be for all assets to go directly into the assets folder of the bucket (which is only used for app assets, and currently contains just folders for every app). I'll amend to simplify. Thanks!

Reply:
Going to mark this as resolved as it's out of scope for the current incarnation of the RFC, but we should definitely revisit RFC-149, because it's a plausible and simple solution for the problem it references that we could implement fairly easily.